sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/RenameTableExec.scala

```diff
@@ -27,14 +27,26 @@ import org.apache.spark.sql.connector.catalog.{Identifier, TableCatalog}
 case class RenameTableExec(
     catalog: TableCatalog,
     oldIdent: Identifier,
-    newIdent: Identifier) extends V2CommandExec {
+    newIdent: Identifier,
+    invalidateCache: () => Option[CacheInfo],
+    cacheTable: (DataSourceV2Relation, CacheInfo) => Unit)
+  extends V2CommandExec {
 
   override def output: Seq[Attribute] = Seq.empty
 
   override protected def run(): Seq[InternalRow] = {
+    import org.apache.spark.sql.connector.catalog.CatalogV2Implicits.IdentifierHelper
+
+    val optOldCacheInfo = invalidateCache()
     catalog.invalidateTable(oldIdent)
+
     catalog.renameTable(oldIdent, newIdent)
 
+    optOldCacheInfo.foreach { cacheInfo =>
+      val tbl = catalog.loadTable(newIdent)
+      val newRelation = DataSourceV2Relation.create(tbl, Some(catalog), Some(newIdent))
+      cacheTable(newRelation, CacheInfo(Some(newIdent.quoted), cacheInfo.storageLevel))
```

Comment:
it's unfortunate that we have to recache the table just for renaming - but yeah I don't know if there is a way to reuse the old in memory relation.

Reply:
If we canonicalize `DataSourceV2Relation`, we may be able to avoid recaching. Canonicalizing `Table` could be tricky and I am not sure if it's worthwhile exploring.